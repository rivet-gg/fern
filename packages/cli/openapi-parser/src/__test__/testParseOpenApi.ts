import { AbsoluteFilePath, join, RelativeFilePath } from "@fern-api/fs-utils";
import { CONSOLE_LOGGER } from "@fern-api/logger";
import { serialization } from "@fern-api/openapi-ir-sdk";
import { createMockTaskContext } from "@fern-api/task-context";
import { parse, SpecImportSettings } from "../parse";

const FIXTURES_PATH = join(AbsoluteFilePath.of(__dirname), RelativeFilePath.of("fixtures"));

// eslint-disable-next-line jest/no-export
export function testParseOpenAPI(
    fixtureName: string,
    openApiFilename: string,
    asyncApiFilename?: string,
    settings?: SpecImportSettings
): void {
    // eslint-disable-next-line jest/valid-title
    describe(fixtureName, () => {
        it("parse open api", async () => {
            const absolutePathToOpenAPI = join(
                FIXTURES_PATH,
                RelativeFilePath.of(fixtureName),
                RelativeFilePath.of(openApiFilename)
            );
            const absolutePathToAsyncAPI =
                asyncApiFilename != null
                    ? join(FIXTURES_PATH, RelativeFilePath.of(fixtureName), RelativeFilePath.of(asyncApiFilename))
                    : undefined;
            const specs = [];
            if (absolutePathToOpenAPI != null) {
                specs.push({
                    absoluteFilepath: absolutePathToOpenAPI,
                    absoluteFilepathToOverrides: undefined,
                    settings
                });
            }
            if (absolutePathToAsyncAPI != null) {
                specs.push({
                    absoluteFilepath: absolutePathToAsyncAPI,
                    absoluteFilepathToOverrides: undefined,
                    settings
                });
            }

            const openApiIr = await parse({
                specs,
                taskContext: createMockTaskContext({ logger: CONSOLE_LOGGER })
            });
            const openApiIrJson = await serialization.OpenApiIntermediateRepresentation.jsonOrThrow(openApiIr, {
                skipValidation: true
            });
            expect(openApiIrJson).toMatchSnapshot();
        });
    });
}
