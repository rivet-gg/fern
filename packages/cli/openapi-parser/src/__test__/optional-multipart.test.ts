import { testParseOpenAPI } from "./testParseOpenApi";

describe("optional-multipart", () => {
    testParseOpenAPI("optional-multipart", "openapi.json");
});
