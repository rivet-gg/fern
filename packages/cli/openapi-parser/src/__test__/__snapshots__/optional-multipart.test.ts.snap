// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`optional-multipart optional-multipart parse open api 1`] = `
{
  "apiVersion": null,
  "basePath": null,
  "channel": [],
  "description": null,
  "endpoints": [
    {
      "audiences": [],
      "authed": false,
      "availability": null,
      "description": "speech",
      "errors": {},
      "examples": [],
      "generatedRequestName": "SpeechOperationRequest",
      "headers": [
        {
          "availability": null,
          "description": "Your API key.",
          "env": null,
          "name": "an-api-key",
          "parameterNameOverride": null,
          "schema": {
            "availability": null,
            "description": "Your API key.",
            "generatedName": "SpeechOperationRequestAnApiKey",
            "groupName": null,
            "nameOverride": null,
            "type": "nullable",
            "value": {
              "availability": null,
              "description": "Your API key.",
              "generatedName": "SpeechOperationRequestAnApiKey",
              "groupName": null,
              "nameOverride": null,
              "schema": {
                "default": null,
                "format": null,
                "maxLength": null,
                "minLength": null,
                "pattern": null,
                "type": "string",
              },
              "type": "primitive",
            },
          },
        },
      ],
      "idempotent": null,
      "internal": null,
      "method": "POST",
      "operationId": "speech_operation",
      "pagination": null,
      "path": "/v1/speech/{voice_id}",
      "pathParameters": [
        {
          "availability": null,
          "description": null,
          "name": "voice_id",
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "SpeechOperationRequestVoiceId",
            "groupName": null,
            "nameOverride": null,
            "schema": {
              "default": null,
              "format": null,
              "maxLength": null,
              "minLength": null,
              "pattern": null,
              "type": "string",
            },
            "type": "primitive",
          },
          "variableReference": null,
        },
      ],
      "queryParameters": [
        {
          "availability": null,
          "description": null,
          "name": "some_query_param",
          "parameterNameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "SpeechOperationRequestSomeQueryParam",
            "groupName": null,
            "nameOverride": null,
            "type": "nullable",
            "value": {
              "availability": null,
              "description": null,
              "generatedName": "SpeechOperationRequestSomeQueryParam",
              "groupName": null,
              "nameOverride": null,
              "schema": {
                "default": 0,
                "exclusiveMaximum": null,
                "exclusiveMinimum": null,
                "maximum": null,
                "minimum": null,
                "multipleOf": null,
                "type": "int",
              },
              "type": "primitive",
            },
          },
        },
      ],
      "request": {
        "description": null,
        "name": "SpeechBody",
        "properties": [
          {
            "description": "The audio file.",
            "key": "audio",
            "schema": {
              "isArray": false,
              "isOptional": false,
              "type": "file",
            },
          },
          {
            "description": null,
            "key": "model_id",
            "schema": {
              "type": "json",
              "value": {
                "availability": null,
                "description": null,
                "generatedName": "speechOperationRequestModelId",
                "groupName": null,
                "nameOverride": "SpeechBody",
                "type": "optional",
                "value": {
                  "availability": null,
                  "description": null,
                  "generatedName": "SpeechOperationRequestModelId",
                  "groupName": null,
                  "nameOverride": null,
                  "schema": {
                    "default": "model-1",
                    "format": null,
                    "maxLength": null,
                    "minLength": null,
                    "pattern": null,
                    "type": "string",
                  },
                  "type": "primitive",
                },
              },
            },
          },
          {
            "description": null,
            "key": "voice_settings",
            "schema": {
              "type": "json",
              "value": {
                "availability": null,
                "description": null,
                "generatedName": "speechOperationRequestVoiceSettings",
                "groupName": null,
                "nameOverride": "SpeechBody",
                "type": "optional",
                "value": {
                  "availability": null,
                  "description": "Voice settings.",
                  "generatedName": "SpeechOperationRequestVoiceSettings",
                  "groupName": null,
                  "nameOverride": null,
                  "schema": {
                    "default": null,
                    "format": null,
                    "maxLength": null,
                    "minLength": null,
                    "pattern": null,
                    "type": "string",
                  },
                  "type": "primitive",
                },
              },
            },
          },
        ],
        "type": "multipart",
      },
      "requestNameOverride": null,
      "response": {
        "description": "Successful Response",
        "type": "file",
      },
      "sdkName": null,
      "server": [],
      "summary": "speech",
      "tags": [],
    },
    {
      "audiences": [],
      "authed": false,
      "availability": null,
      "description": "speech_stream",
      "errors": {},
      "examples": [
        {
          "codeSamples": [],
          "description": null,
          "headers": [],
          "name": null,
          "pathParameters": [
            {
              "name": "voice_id",
              "value": {
                "type": "primitive",
                "value": {
                  "type": "string",
                  "value": "1234",
                },
              },
            },
          ],
          "queryParameters": [],
          "request": {
            "properties": {},
            "type": "object",
          },
          "response": null,
          "type": "full",
        },
      ],
      "generatedRequestName": "SpeechStreamOperationRequest",
      "headers": [
        {
          "availability": null,
          "description": "Your API key.",
          "env": null,
          "name": "an-api-key",
          "parameterNameOverride": null,
          "schema": {
            "availability": null,
            "description": "Your API key.",
            "generatedName": "SpeechStreamOperationRequestAnApiKey",
            "groupName": null,
            "nameOverride": null,
            "type": "nullable",
            "value": {
              "availability": null,
              "description": "Your API key.",
              "generatedName": "SpeechStreamOperationRequestAnApiKey",
              "groupName": null,
              "nameOverride": null,
              "schema": {
                "default": null,
                "format": null,
                "maxLength": null,
                "minLength": null,
                "pattern": null,
                "type": "string",
              },
              "type": "primitive",
            },
          },
        },
      ],
      "idempotent": null,
      "internal": null,
      "method": "POST",
      "operationId": "speech_stream_operation",
      "pagination": null,
      "path": "/v1/speech/{voice_id}/stream",
      "pathParameters": [
        {
          "availability": null,
          "description": null,
          "name": "voice_id",
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "SpeechStreamOperationRequestVoiceId",
            "groupName": null,
            "nameOverride": null,
            "schema": {
              "default": null,
              "format": null,
              "maxLength": null,
              "minLength": null,
              "pattern": null,
              "type": "string",
            },
            "type": "primitive",
          },
          "variableReference": null,
        },
      ],
      "queryParameters": [
        {
          "availability": null,
          "description": null,
          "name": "some_query_param",
          "parameterNameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "SpeechStreamOperationRequestSomeQueryParam",
            "groupName": null,
            "nameOverride": null,
            "type": "nullable",
            "value": {
              "availability": null,
              "description": null,
              "generatedName": "SpeechStreamOperationRequestSomeQueryParam",
              "groupName": null,
              "nameOverride": null,
              "schema": {
                "default": 0,
                "exclusiveMaximum": null,
                "exclusiveMinimum": null,
                "maximum": null,
                "minimum": null,
                "multipleOf": null,
                "type": "int",
              },
              "type": "primitive",
            },
          },
        },
      ],
      "request": {
        "description": null,
        "name": "SpeechBodyStream",
        "properties": [
          {
            "description": "The audio file.",
            "key": "audio",
            "schema": {
              "isArray": false,
              "isOptional": false,
              "type": "file",
            },
          },
          {
            "description": null,
            "key": "model_id",
            "schema": {
              "type": "json",
              "value": {
                "availability": null,
                "description": null,
                "generatedName": "speechStreamOperationRequestModelId",
                "groupName": null,
                "nameOverride": "SpeechBodyStream",
                "type": "optional",
                "value": {
                  "availability": null,
                  "description": null,
                  "generatedName": "SpeechStreamOperationRequestModelId",
                  "groupName": null,
                  "nameOverride": null,
                  "schema": {
                    "default": "model-1",
                    "format": null,
                    "maxLength": null,
                    "minLength": null,
                    "pattern": null,
                    "type": "string",
                  },
                  "type": "primitive",
                },
              },
            },
          },
          {
            "description": null,
            "key": "voice_settings",
            "schema": {
              "type": "json",
              "value": {
                "availability": null,
                "description": null,
                "generatedName": "speechStreamOperationRequestVoiceSettings",
                "groupName": null,
                "nameOverride": "SpeechBodyStream",
                "type": "optional",
                "value": {
                  "availability": null,
                  "description": "Voice settings.",
                  "generatedName": "SpeechStreamOperationRequestVoiceSettings",
                  "groupName": null,
                  "nameOverride": null,
                  "schema": {
                    "default": null,
                    "format": null,
                    "maxLength": null,
                    "minLength": null,
                    "pattern": null,
                    "type": "string",
                  },
                  "type": "primitive",
                },
              },
            },
          },
          {
            "description": null,
            "key": "my_enum",
            "schema": {
              "type": "json",
              "value": {
                "availability": null,
                "description": null,
                "generatedName": "speechStreamOperationRequestMyEnum",
                "groupName": null,
                "nameOverride": "SpeechBodyStream",
                "type": "optional",
                "value": {
                  "availability": null,
                  "default": null,
                  "description": null,
                  "generatedName": "SpeechStreamOperationRequestMyEnum",
                  "groupName": null,
                  "nameOverride": null,
                  "type": "enum",
                  "values": [
                    {
                      "availability": null,
                      "casing": {
                        "camel": null,
                        "pascal": null,
                        "screamingSnake": null,
                        "snake": null,
                      },
                      "description": null,
                      "generatedName": "default",
                      "nameOverride": null,
                      "value": "default",
                    },
                    {
                      "availability": null,
                      "casing": {
                        "camel": null,
                        "pascal": null,
                        "screamingSnake": null,
                        "snake": null,
                      },
                      "description": null,
                      "generatedName": "converting",
                      "nameOverride": null,
                      "value": "converting",
                    },
                  ],
                },
              },
            },
          },
          {
            "description": null,
            "key": "my_ref",
            "schema": {
              "type": "json",
              "value": {
                "availability": null,
                "description": null,
                "generatedName": "speechStreamOperationRequestMyRef",
                "groupName": null,
                "nameOverride": "SpeechBodyStream",
                "type": "optional",
                "value": {
                  "availability": null,
                  "description": null,
                  "generatedName": "SpeechStreamOperationRequestMyRef",
                  "groupName": null,
                  "nameOverride": null,
                  "schema": "request-model",
                  "type": "reference",
                },
              },
            },
          },
        ],
        "type": "multipart",
      },
      "requestNameOverride": null,
      "response": null,
      "sdkName": null,
      "server": [],
      "summary": "speech_stream",
      "tags": [],
    },
  ],
  "globalHeaders": [],
  "groups": {},
  "hasEndpointsMarkedInternal": false,
  "idempotencyHeaders": [],
  "nonRequestReferencedSchemas": [
    "request-model",
  ],
  "schemas": {
    "SpeechBody": {
      "additionalProperties": false,
      "allOf": [],
      "allOfPropertyConflicts": [],
      "availability": null,
      "description": null,
      "generatedName": "SpeechBody",
      "groupName": null,
      "nameOverride": "SpeechBody",
      "properties": [
        {
          "audiences": [],
          "availability": null,
          "conflict": {},
          "generatedName": "speechBodyAudio",
          "key": "audio",
          "nameOverride": null,
          "schema": {
            "availability": null,
            "description": "The audio file.",
            "generatedName": "SpeechBodyAudio",
            "groupName": null,
            "nameOverride": "Audio",
            "schema": {
              "default": null,
              "format": "binary",
              "maxLength": null,
              "minLength": null,
              "pattern": null,
              "type": "string",
            },
            "type": "primitive",
          },
        },
        {
          "audiences": [],
          "availability": null,
          "conflict": {},
          "generatedName": "speechBodyModelId",
          "key": "model_id",
          "nameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "speechBodyModelId",
            "groupName": null,
            "nameOverride": "SpeechBody",
            "type": "optional",
            "value": {
              "availability": null,
              "description": null,
              "generatedName": "SpeechBodyModelId",
              "groupName": null,
              "nameOverride": null,
              "schema": {
                "default": "model-1",
                "format": null,
                "maxLength": null,
                "minLength": null,
                "pattern": null,
                "type": "string",
              },
              "type": "primitive",
            },
          },
        },
        {
          "audiences": [],
          "availability": null,
          "conflict": {},
          "generatedName": "speechBodyVoiceSettings",
          "key": "voice_settings",
          "nameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "speechBodyVoiceSettings",
            "groupName": null,
            "nameOverride": "SpeechBody",
            "type": "optional",
            "value": {
              "availability": null,
              "description": "Voice settings.",
              "generatedName": "SpeechBodyVoiceSettings",
              "groupName": null,
              "nameOverride": null,
              "schema": {
                "default": null,
                "format": null,
                "maxLength": null,
                "minLength": null,
                "pattern": null,
                "type": "string",
              },
              "type": "primitive",
            },
          },
        },
      ],
      "type": "object",
    },
    "SpeechBodyStream": {
      "additionalProperties": false,
      "allOf": [],
      "allOfPropertyConflicts": [],
      "availability": null,
      "description": null,
      "generatedName": "SpeechBodyStream",
      "groupName": null,
      "nameOverride": "SpeechBodyStream",
      "properties": [
        {
          "audiences": [],
          "availability": null,
          "conflict": {},
          "generatedName": "speechBodyStreamAudio",
          "key": "audio",
          "nameOverride": null,
          "schema": {
            "availability": null,
            "description": "The audio file.",
            "generatedName": "SpeechBodyStreamAudio",
            "groupName": null,
            "nameOverride": "Audio",
            "schema": {
              "default": null,
              "format": "binary",
              "maxLength": null,
              "minLength": null,
              "pattern": null,
              "type": "string",
            },
            "type": "primitive",
          },
        },
        {
          "audiences": [],
          "availability": null,
          "conflict": {},
          "generatedName": "speechBodyStreamModelId",
          "key": "model_id",
          "nameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "speechBodyStreamModelId",
            "groupName": null,
            "nameOverride": "SpeechBodyStream",
            "type": "optional",
            "value": {
              "availability": null,
              "description": null,
              "generatedName": "SpeechBodyStreamModelId",
              "groupName": null,
              "nameOverride": null,
              "schema": {
                "default": "model-1",
                "format": null,
                "maxLength": null,
                "minLength": null,
                "pattern": null,
                "type": "string",
              },
              "type": "primitive",
            },
          },
        },
        {
          "audiences": [],
          "availability": null,
          "conflict": {},
          "generatedName": "speechBodyStreamVoiceSettings",
          "key": "voice_settings",
          "nameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "speechBodyStreamVoiceSettings",
            "groupName": null,
            "nameOverride": "SpeechBodyStream",
            "type": "optional",
            "value": {
              "availability": null,
              "description": "Voice settings.",
              "generatedName": "SpeechBodyStreamVoiceSettings",
              "groupName": null,
              "nameOverride": null,
              "schema": {
                "default": null,
                "format": null,
                "maxLength": null,
                "minLength": null,
                "pattern": null,
                "type": "string",
              },
              "type": "primitive",
            },
          },
        },
        {
          "audiences": [],
          "availability": null,
          "conflict": {},
          "generatedName": "speechBodyStreamMyEnum",
          "key": "my_enum",
          "nameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "speechBodyStreamMyEnum",
            "groupName": null,
            "nameOverride": "SpeechBodyStream",
            "type": "optional",
            "value": {
              "availability": null,
              "default": null,
              "description": null,
              "generatedName": "SpeechBodyStreamMyEnum",
              "groupName": null,
              "nameOverride": null,
              "type": "enum",
              "values": [
                {
                  "availability": null,
                  "casing": {
                    "camel": null,
                    "pascal": null,
                    "screamingSnake": null,
                    "snake": null,
                  },
                  "description": null,
                  "generatedName": "default",
                  "nameOverride": null,
                  "value": "default",
                },
                {
                  "availability": null,
                  "casing": {
                    "camel": null,
                    "pascal": null,
                    "screamingSnake": null,
                    "snake": null,
                  },
                  "description": null,
                  "generatedName": "converting",
                  "nameOverride": null,
                  "value": "converting",
                },
              ],
            },
          },
        },
        {
          "audiences": [],
          "availability": null,
          "conflict": {},
          "generatedName": "speechBodyStreamMyRef",
          "key": "my_ref",
          "nameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "speechBodyStreamMyRef",
            "groupName": null,
            "nameOverride": "SpeechBodyStream",
            "type": "optional",
            "value": {
              "availability": null,
              "description": null,
              "generatedName": "SpeechBodyStreamMyRef",
              "groupName": null,
              "nameOverride": null,
              "schema": "request-model",
              "type": "reference",
            },
          },
        },
      ],
      "type": "object",
    },
    "request-model": {
      "additionalProperties": false,
      "allOf": [],
      "allOfPropertyConflicts": [],
      "availability": null,
      "description": null,
      "generatedName": "RequestModel",
      "groupName": null,
      "nameOverride": null,
      "properties": [
        {
          "audiences": [],
          "availability": null,
          "conflict": {},
          "generatedName": "requestModelVoiceId",
          "key": "voice_id",
          "nameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "RequestModelVoiceId",
            "groupName": null,
            "nameOverride": null,
            "schema": {
              "default": null,
              "format": null,
              "maxLength": null,
              "minLength": null,
              "pattern": null,
              "type": "string",
            },
            "type": "primitive",
          },
        },
      ],
      "type": "object",
    },
  },
  "securitySchemes": {},
  "servers": [],
  "tags": {
    "orderedTagIds": null,
    "tagsById": {},
  },
  "title": "API Documentation",
  "variables": {},
  "webhooks": [],
}
`;
